contracts/Divvi.sol

```diff
@@ -12,17 +12,58 @@ contract DivviToken is
 {
   bytes32 public constant MINTER_ROLE = keccak256('MINTER_ROLE');
 
+  mapping(address => bool) private _hasTransferPermission;
+  bool private _permitTransfersOnly;
+
+  error SenderNotPermitted();
+
   function initialize(address owner, uint48 transferDelay) public initializer {
     __ERC20_init('Divvi', 'DIVVI');
     __ERC20Permit_init('Divvi');
     __AccessControlDefaultAdminRules_init(transferDelay, owner);
     __UUPSUpgradeable_init();
+
+    _permitTransfersOnly = true;
   }
 
   function mint(address to, uint256 amount) external onlyRole(MINTER_ROLE) {
     _mint(to, amount);
   }
 
+  function transfer(
+    address recipient,
+    uint256 amount
+  ) public override returns (bool) {
+    if (!isPermittedToTransfer(msg.sender)) revert SenderNotPermitted();
+    return super.transfer(recipient, amount);
+  }
+
+  function transferFrom(
+    address sender,
+    address recipient,
+    uint256 amount
+  ) public override returns (bool) {
+    if (!isPermittedToTransfer(sender)) revert SenderNotPermitted();
+    return super.transferFrom(sender, recipient, amount);
+  }
+
+  function setPermitTransfersOnly(
+    bool status
+  ) external onlyRole(DEFAULT_ADMIN_ROLE) {
+    _permitTransfersOnly = status;
+  }
+
+  function setPermittedSender(
```

Comment:
naming nit, maybe `setSenderPermission` or something? the naming kind of implies this is just used to permit someone to send, but it in fact is used in both directions.

Reply:
makes sense, updated!